Python/context.c

```diff
@@ -125,11 +125,14 @@ notify_context_watchers(PyThreadState *ts, PyContextEvent event, PyObject *ctx)
         if (bits & 1) {
             PyContext_WatchCallback cb = interp->context_watchers[i];
             assert(cb != NULL);
-            if (cb(event, ctx) < 0) {
+            PyObject *exc = _PyErr_GetRaisedException(ts);
+            cb(event, ctx);
+            if (_PyErr_Occurred(ts) != NULL) {
```

Comment:
Can this use ``cb``'s return value as before, rather than checking ``_PyErr_Occurred``?

Reply:
This PR changes the callback's return type to `void`  to discourage the use of exceptions (since they get ignored anyway), so there is no return value to check.

Reply:
I noticed, but I'm suggesting to not do that. 

Also, the documentation is incorrect: `"If the callback raises an exception it will be ignored."`
 it's not completely ignored.

Reply:
Ah, I see.  I think that discarding this PR's second commit would address both of your concerns, though I would still like to tweak the documentation's wording a bit to be more precise.

Part of the reason I want to change the return value to `void` is to cement a future where context switches cannot fail.  For gh-99633, context switches can happen when generators or coroutines resume or suspend, where a failure to switch would be particularly difficult to handle correctly.  I want the event notification API to make it clear to developers that the callback is informational only; the context switch will happen even if they don't want it to.

(I should express this rationale in the commit message and PR description.)

Reply:
I tweaked the wording of the documentation and the commit message.  I still believe that it is better to return `void` so I kept that, but I can easily undo it by discarding the final commit if you disagree.  Please let me know what you think.

Reply:
I see your point.

Reply:
@iritkatriel 
> Can this use cb's return value as before, rather than checking _PyErr_Occurred?
> I noticed, but I'm suggesting to not do that.

Why using `_PyErr_Occurred` is a bad idea?

Reply:
If you use the return value then you know that the error is from the function you just called. If you use ``_PyErr_Occurred``, the exception could have been there before that call, so it's less specific.

Reply:
But in this case we know where the exception is coming from, `_PyErr_GetRaisedException` resets the current exception to `NULL`. So if `_PyErr_Occurred` returns anything it mush be coming from the callback.

Reply:
(FWIW I don't feel strong about this, but I do like the idea somewhat)

Reply:
@iritkatriel Does the thumbs-up reaction on @1st1's comment mean you're OK with this PR?

Reply:
@ambv @pablogsal can one of you weigh in here and give a third opinion? I basically like the PR as is as I think it makes a better API this way, but I also understand the point Irit is making. Also this might be the first CPython API designed this way.